nistats/reporting/glm_reporter.py

```diff
@@ -0,0 +1,561 @@
+                               )
+from nistats.thresholding import map_threshold
+
+
+html_template_root_path = os.path.dirname(__file__)
+
+
+def make_glm_report(
+        model,
+        contrasts,
+        title='auto',
+        roi_img=None,
+        bg_img='MNI 152 Template',
+        threshold=3.09,
+        alpha=0.01,
+        cluster_threshold=None,
+        height_control='fpr',
+        min_distance=8.,
+        plot_type='stat',
+        display_mode=None,
+        cut_coords=None,
+        nb_width=1600,
+        nb_height=800,
+        ):
+    """ Returns HTMLDocument object for a report which shows
+    all important aspects of a fitted GLM.
+    The object can be opened in a browser or saved to disk.
+    
+    Usage:
+        report = make_glm_report(model, contrasts)
+        report.open_in_browser()
+        report.save_as_html(destination_path)
+    
+    Parameters
+    ----------
+    model: FirstLevelModel or SecondLevelModel object
+        A fitted first or second level model object.
+        
+    contrasts: Dict[string, ndarray] , String
+    
+    title: str, default 'auto'
+        Text to represnt the web page's title and primary heading.
+        
+    threshold: float
+        Cluster forming threshold in same scale as `stat_img` (either a
+        p-value or z-scale value).
+        Default is 3.09
+        
+    alpha: float
+        P- value for clustering.
+
+    cluster_threshold : int or None, optional
+        Cluster size threshold, in voxels.
+        
+    height_control: str
+
+    min_distance: float, optional
+        Minimum distance between subpeaks in mm. Default is 8 mm.
+        
+    bg_img: Nifti image
+        Default is the MNI152 template
+        
+    display_mode: String, optional
+        Default is 'z' if plot_type is 'stat'; 'ortho' if plot_type is 'glass'.
+        
+    plot_type: String. ['stat' (default)| 'glass']
+        Specifies the type of plot to be drawn for the statistical maps.
+        
+    Returns
+    -------
+    report_text: HTMLDocument Object
+        Contains the HTML code for the GLM Report.
+    """
+    if not display_mode:
+        if plot_type == 'stat':
+            display_mode = 'z'
+        elif plot_type == 'glass':
+            display_mode = 'lzry'
+    cluster_threshold = cluster_threshold if cluster_threshold else 0
+    pd.set_option('display.max_colwidth', -1)
+    bg_img = load_mni152_template() if bg_img == 'MNI 152 Template' else bg_img
+    html_template_path = os.path.join(html_template_root_path,
+                                      'report_template.html')
+    with open(html_template_path) as html_file_obj:
+        html_template_text = html_file_obj.read()
+    report_template = string.Template(html_template_text)
+    
+    contrasts = _make_contrasts_dict(contrasts)
+    contrast_plots = _make_dict_of_contrast_plots(contrasts, model)
+    all_contrast_plots_html = ['{}<p>{}'.format(key, item)
```

Comment:
I think it would be best to avoid these string manipulations inside the python code to write the html, and use either a templating engine or a library dedicated to writing html documents.

the contrast names, provided by the  user, should be escaped. For example it is common to describe contrasts as "something > something else" but ">" is not a valid character in html.

each contrast should probably be in its own div rather than a flat sequence of paragraphs alternating contrast name and plot

Reply:
Unnecessarily adding `<divs>` is a bad practice. Since we are not applying any CSS, there is no benefit to having divs all over the place.
I am working on removing any HTML I have added in the python code.
I was under the impression that a '<' or '>' by itself doesn't mean anything; it has to be a proper tag for it to interfere with the HTML.

Reply:
By proper tag I mean anything in the format `<text>`

Reply:
> I was under the impression that a '<' or '>' by itself doesn't mean anything; it has to be a proper tag for it to interfere with the HTML.

no, `<` and `>` are reserved characters and must be replaced with character entities `&lt;`, `&gt;`

Reply:
Currently, contrast labels are only being used in one place, to give a title to every set of stat map + contrast plot + cluster table, and it is being rendered correctly.
![image](https://user-images.githubusercontent.com/8203204/62290365-83f1cc80-b461-11e9-94d8-10f70a27920e.png)
Unless there's a specific use case where this is raising an error (and we would need one to write the code & test), I think this is fine.
Are you concerned this will render the HTML non-compliant?

Reply:
Eh. I have fixed it anyway.

Reply:
thanks. browsers are excellent at coping with broken html but we should still strive to have valid reports